### ISSUE TITLE
Error Probagation

### ISSUE THREAD

**Issue by KhoiDOO** [sail-sg/TreeMeshGPT#10]
Hi authors,

I just wonder how you can figure out in case error propagation happens. For example, when some vertices or some coordinates are missing during generation, then the mesh cannot be reconstructed. 

Thank you!

**Comment by splionar** [comment]
Hi @KhoiDOO ,

We do not have any specific strategy to minimize error propagation in this work. Error accumulation and hallucination are common issues for autoregressive models. Recent works have proposed post-training alignment using reinforcement learning (https://arxiv.org/abs/2503.15265 & https://arxiv.org/abs/2505.16761). I think it can be a promising direction to minimize error accumulation.

**Comment by KhoiDOO** [comment]
Thank you so much @splionar, for your information.

I'm trying to retrain the model with the processed data (following your pipeline) that I downloaded from Objarverse. Now I have converted them all to sequence and sampled their point cloud too. 

May I ask the best loss value in training/validation that you achieved during training the model, or any other things that I may consider? That will be good information for me to make sure that I'm going the right way.

Thank you!

**Comment by splionar** [comment]
Here is our train/val CE loss curve. It took around 5 days to reach 80k steps (loss is about 0.53 train and 0.55 val), then we continued training until 140k steps for the released checkpoint.

![Image](https://github.com/user-attachments/assets/6b16f1f2-931f-488f-8759-80b8f41ffea2)
![Image](https://github.com/user-attachments/assets/ee8272ad-195d-4808-a247-9182b46e7334)

**Comment by KhoiDOO** [comment]
Thank you so much for your prompt response. 

I'm just curious if there is a reason behind eos_token = -2? In training Language model eos_token is usually ```vocab_size```, did you use ```vocab_size``` for the other token? Otherwise, I did not find where you defined the sos_token. Is it pre-defined or a learnable token? 

Thank you!

**Comment by splionar** [comment]
https://github.com/sail-sg/TreeMeshGPT/blob/35821582ac14090f2a9cb173e4ea5fa022ca760a/model/treemeshgpt_train.py#L135

@KhoiDOO If you check further down the code, it is mapped to the vocabulary size (depending on the quantization bit).

https://github.com/sail-sg/TreeMeshGPT/blob/35821582ac14090f2a9cb173e4ea5fa022ca760a/model/treemeshgpt_train.py#L144

Here is where the start tokens for each connected component are handled.

**Comment by KhoiDOO** [comment]
Thank you so much for your clarification, @splionar 

I have two more questions
- I saw that in [train_create_pkl.py](https://github.com/sail-sg/TreeMeshGPT/blob/main/train_create_pkl.py) you perform an infinite loop to sample multiple augmented samples, is it similar if I integrate online augmentation into the dataset?
- Do you have training/model settings for the mesh_500.zip that you provided?

Thank you so so much!

**Comment by splionar** [comment]
@KhoiDOO 

I initially implemented the augmentation and sequencing in PyTorch dataloader, but encountered issue with Open3D half-edge functions with num_workers > 0.

The mesh_500.zip in the our experiment is to ablate with other methods when trained from scratch. All training configurations are the same, except model size is smaller (16-layer Transformers with
768 hidden dimension)

**Comment by KhoiDOO** [comment]
Once again, thank you so much, @splionar, for your help. I will close the issue here as I currently do not have other questions.
